# A set of random scripts

Random scripts to deal with common fasta files, phylogenetic trees, tab-delimited lists and other fancy stuff. Such as split an alignment in defined positions, concatenate different fasta files, extract/remove sequences, reorder an alignment based on a tree, get the entropy on each position from an alignment, get the consensus sequence from an alignment, ...

Download and move the scripts to you prefered folder (e.g.;```/usr/lobal/bin/```) and start using them (you might have to make them executable: ```chmod +x *.py```).  
  
## A brief description of most useful scripts
  
### For fasta files manipulations    
  
*alignmentConsensus.py*: Creates a consensus of an aligned fasta file.  
alignmentEntropy.py: Exports a table with the entropy and other values for every position of an aligned fasta file.  
fasta2phylip.py
fastaChangeCases.py
fastaClean.py
fastaConcat.py
fastaRarefy.py
fastaRename.py
fastaRevCom.py
fastaSplit.py
fastaStats.py
multi2line
multi2linefasta.py
phylip2fasta.py
sequenceSelect.py  
  
Further details in the help of each script (```script -h```).

Please feel free to report bugs or suggestions.
